engine/table/src/main/java/io/deephaven/engine/table/impl/BlinkTableTools.java

```diff
@@ -183,4 +108,133 @@ public static boolean isBlink(Table table) {
+                final WritableColumnSource<?> newColumn = ArrayBackedColumnSource.getMemoryColumnSource(
+                        0, existingColumn.getType(), existingColumn.getComponentType());
+                resultColumns.put(nameColumnSourceEntry.getKey(), newColumn);
+
+                // read and write primitives whenever possible
+                sourceColumns[colIdx] = ReinterpretUtils.maybeConvertToPrimitive(existingColumn);
+                destColumns[colIdx++] = ReinterpretUtils.maybeConvertToWritablePrimitive(newColumn);
+            }
+
+            final TrackingWritableRowSet rowSet = RowSetFactory.empty().toTracking();
+            resultTable = new QueryTable(rowSet, resultColumns);
+            resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+            resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+            resultTable.setFlat();
+
+            // statistically a blink table is likely to be empty; but let's copy whatever we can
+            appendRows(usePrev ? parent.getRowSet().prev() : parent.getRowSet(), usePrev).close();
+            rowSet.initializePreviousValue();
+            Assert.leq(rowSet.size(), "rowSet.size()", sizeLimit, "sizeLimit");
+
+            if (resultTable.size() < sizeLimit) {
+                resultTable.setRefreshing(true);
+
+                resultListener = new BaseTable.ListenerImpl(getDescription(), parent, resultTable) {
+                    @Override
+                    public void onUpdate(TableUpdate upstream) {
+                        BlinkToAppendOnlyOperation.this.onUpdate(upstream);
+                    }
+                };
+            }
+
+            return new Result<>(resultTable, resultListener);
+        }
+
+        private void onUpdate(final TableUpdate upstream) {
+            if (upstream.modified().isNonempty() || upstream.shifted().nonempty()) {
+                throw new IllegalStateException("Blink tables should not modify or shift!");
+            }
+            long newRowsSize = upstream.added().size();
+            if (newRowsSize == 0) {
+                return;
+            }
+
+            final TableUpdateImpl downstream = new TableUpdateImpl();
+            downstream.added = appendRows(upstream.added(), false);
+            downstream.modified = RowSetFactory.empty();
+            downstream.removed = RowSetFactory.empty();
+            downstream.modifiedColumnSet = ModifiedColumnSet.EMPTY;
+            downstream.shifted = RowSetShiftData.EMPTY;
+            resultTable.notifyListeners(downstream);
+
+            if (resultTable.size() == sizeLimit) {
+                // No more rows can be appended, so remove the listener and remove all references
+                resultListener.forceReferenceCountToZero();
+                resultListener = null;
+            }
+        }
+
+        private RowSet appendRows(final RowSet newRows, final boolean usePrev) {
```

Comment:
* This could easily accept the `currentSize` as an argument.
* That would let you defer creating the result table until after the `copyData`.
* That, in turn, would let you avoid making the result row set tracking prematurely, avoid the need to use `initializePreviousValue` (which is almost always a code smell), and let you always use the returned `newRange`, because you can just convert it to tracking to be the initial row set, and insert it *outside* of this method in the updating case.
* (Note that you can do the `insertRange` using `newRange.firstRowKey()` and `newRange.lastRowKey()`.)